gateways/core/artifact/s3.go

```diff
@@ -40,18 +40,24 @@ const (
 
 type s3 struct {
 	// gatewayConfig provides a generic configuration for a gateway
-	gatewayConfig *gateways.GatewayConfig
-	registeredArtifacts []S3Artifact
+	gatewayConfig       *gateways.GatewayConfig
+	// registeredArtifacts contains map of registered s3 artifacts
+	registeredArtifacts map[uint64]*S3Artifact
```

Comment:
what is the purpose of this map? it doesn't look like we're saving anything in this? also the gateway doesn't want to be storing state correct?

Reply:
A gateway can be configured with multiple configuration. Each configuration is run in different go routine. The idea is to use a single gateway among multiple use cases that differ only in configurations. 

We are watching gateway-configmap to dynamically add new configuration and the purpose of this map is to keep track of configurations of gateway. Whenever an user updates the configmap with new configuration, this map is used to first check whether the configuration already exists or is it a completely new configuration.

Reply:
so what happens when a gateway goes down and loses this information? does it pick up the existing configurations from the configMap and restart the necessary go routines?